src/engine/sidechain/engineshoutcast.cpp

```diff
@@ -787,61 +719,40 @@ void EngineShoutcast::infoDialog(QString text, QString detailedInfo) {
-    while (1)
-    {
-       sleep(5);
-       m_SMutex.lock();
-       m_lCacheSize = m_pShoutcastCache.size();
-       if (m_lCacheSize > 10) {
-          m_SMutex.unlock();
-          break;
-       }
-       m_SMutex.unlock();
-    }
-
-    while (1)
-    {
-        shout_sync(m_pShout);
-        m_SMutex.lock();
-        // Ok this shouldn't happen cache is empty and we have nothing to send.
-        // wait for next to come.
-        if (m_bThreadQuit == false && m_pShoutcastCache.size() == 0) {
-            qDebug() << "EngineShoutcast::shoutcastThread: Encoding not fast enough. Cache is empty!";
-            qDebug() << "EngineShoutcast::shoutcastThread: This means your stream will break";
-            m_SMutex.unlock();
-            sleep(5);
-            continue;
-        } else if (m_pShoutcastCache.size() > 0) {
-            // Take first cache object and remove it
-            l_SCacheObj = m_pShoutcastCache.first();
-            m_pShoutcastCache.remove(0);
-            m_lCacheSize = m_pShoutcastCache.size();
-        } else {
-            // We don't have anything to send anymore
-            l_SCacheObj = NULL;
-            m_lCacheSize = 0;
-        }
-        m_SMutex.unlock();
-
-        if (l_SCacheObj != NULL) {
-            // I don't know if emiting Signal is more QT style but doing it like this now.
-            serverWrite(l_SCacheObj->header,
-                        l_SCacheObj->body,
-                        l_SCacheObj->headerLen,
-                        l_SCacheObj->bodyLen);
-
-            free(l_SCacheObj->header);
-            free(l_SCacheObj->body);
-            free(l_SCacheObj);
-        }
+// Is called from the Mixxx engine thread
+void EngineShoutcast::outputAvailabe(FIFO<CSAMPLE>* pOutputFifo) {
+    m_pOutputFifo = pOutputFifo;
+    m_readSema.release();
+}
 
-        if (m_bThreadQuit == true && m_lCacheSize == 0) {
-          break;
+void EngineShoutcast::run() {
+    unsigned static id = 0;
+    QThread::currentThread()->setObjectName(QString("EngineShoutcast %1").arg(++id));
+    qDebug() << "starting thread";
+    for(;;) {
+        m_readSema.acquire();
+        if (m_bThreadQuit) {
+            if (m_encoder) {
+                m_encoder->flush();
+                delete m_encoder;
+                m_encoder = NULL;
+            }
+            return;
+        }
+        int readAvailable = m_pOutputFifo->readAvailable();
+        if (readAvailable) {
+            CSAMPLE* dataPtr1;
+            ring_buffer_size_t size1;
+            CSAMPLE* dataPtr2;
+            ring_buffer_size_t size2;
+            // We use size1 and size2, so we can ignore the return value
+            (void)m_pOutputFifo->aquireReadRegions(readAvailable, &dataPtr1, &size1,
+                    &dataPtr2, &size2);
+            process(dataPtr1, size1);
+            if (size2 > 0) {
+                process(dataPtr1, size2);
+            }
+            m_pOutputFifo->releaseReadRegions(readAvailable);
```

Comment:
One critical point is here: we use releaseReadRegions here from the shoutcast thread and from the engine theread when we flush in case the buffer is full. This is a serious race condition. A mutex could be used to guard it. but we worked hard to keep the Mixxx engine lock free. ... Any ideas?